fix(http): queue jsonp <script> tag onLoad event handler in microtask

sebastianhaeni opened the pull request:
Fixes #39496

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #39496

Before this change, when trying to load a JSONP script that calls the JSONP callback inside a setTimeout function, it will fail in Internet Explorer 11 and EdgeHTML.

## What is the new behavior?

This commit changes the event handler on the script tag to be queued with setTimeout. This ensures that the aforementioned browsers will first evaluate the loaded script calling the JSONP callback and only then run the load handler which does the tear down of the callback function and script tag.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

sebastianhaeni committed:
fix(http): queue jsonp <script> tag onLoad event handler in microtask

Before this change, when trying to load a JSONP script that calls the JSONP callback inside a
microtask, it will fail in Internet Explorer 11 and EdgeHTML. This commit changes the onLoad cleanup
to be queued after the loaded endpoint executed any potential microtask itself. This ensures that
the aforementioned browsers will first evaluate the loaded script calling the JSONP callback and
only then run the cleanup inside onLoad.

Fixes #39496

JiaLiPassion reviewed (changes requested):
just like I commented in https://github.com/angular/angular/issues/39496, I think we can use a `promise.then` here to delay the execution. 

Also could you add a test case to verify this fix?

sebastianhaeni commented:
@JiaLiPassion thank you for your comments!
I updated the PR with your proposed fix and added a test. I also verified this solution works in `IE11`, and it does.
Please have a look again.

JiaLiPassion commented on packages/common/http/src/jsonp.ts:
```diff
@@ -189,9 +189,12 @@ export class JsonpClientBackend implements HttpBackend {
+      // and add it to the page. We wrap it in an extra Promise, to ensure the microtask
+      // is scheduled after the loaded endpoint has executed any potential microtask itself,
+      // which is not guaranteed in Internet Explorer and EdgeHTML.
+      node.addEventListener('load', (event: Event) => Promise.resolve().then(() => onLoad(event)));
```
We can use one `resolved promise` instance here, so we don't need to create a new resolved promise every time.

sebastianhaeni replied:
Thanks for this suggestion. I will create a lazy initialized class member with a resolved promise. 

After reading through the code again, I noticed that, the way I did it, `removeEventListener` would break. That's why I'm now invoking `then()` inside the `onLoad` function and I completely removed any change to `onError` as it does not need to be queued after the loaded code has executed any potential microtask.

JiaLiPassion commented on packages/common/http/src/jsonp.ts:
```diff
@@ -189,9 +189,12 @@ export class JsonpClientBackend implements HttpBackend {
+      // is scheduled after the loaded endpoint has executed any potential microtask itself,
+      // which is not guaranteed in Internet Explorer and EdgeHTML.
+      node.addEventListener('load', (event: Event) => Promise.resolve().then(() => onLoad(event)));
+      node.addEventListener(
```
The same with above, we can use one promise here.

sebastianhaeni replied:
I will revert the change to `onError` as it does not need to be queued. The error event is triggered if the resource could not be loaded (e.g. a failed network request). Thus there won't be any inside promise to be resolved before we handle the error.

JiaLiPassion commented on packages/common/http/test/jsonp_spec.ts:
```diff
@@ -45,6 +45,15 @@ const SAMPLE_REQ = new HttpRequest<never>('JSONP', '/test');
     });
+    it('handles a request with callback call wrapped in promise', done => {
+      backend.handle(SAMPLE_REQ).subscribe(() => {
+        done();
```
I didn't test it, but if we write `done()` here, the test case will pass even before this PR? And the `runOnlyCallback` will throw error after `done()`? 
Could you confirm it?

sebastianhaeni replied:
With the old implementation, the test does fail.
That's because the execution order in this test with the old implementation is: `mockLoad -> handle request -> onLoad -> runOnlyCallback`. This means it will fail in `runOnlyCallback` because the function on `home` will not exist anymore.
The new sequence is: `mockLoad -> handle request (queue onLoad) -> runOnlyCallback -> onLoad`.

It outputs this with the old implementation:
```
ERROR: 'Unhandled Promise rejection:', 'callback is not a function', '; Zone:', 'ProxyZone', '; Task:', 'Promise.then', '; Value:', TypeError: callback is not a function
TypeError: callback is not a function
```
```
HeadlessChrome 84.0.4147 (Mac OS X 10.15.7) JsonpClientBackend handles a request with callback call wrapped in promise FAILED
	Error: Expected 0 to be 1.
	    at <Jasmine>
	    at runOnlyCallback (packages/common/http/test/jsonp_spec.ts:19:23 <- angular/packages/common/http/test/jsonp_spec.js:27:29)
	    at ../../packages/common/http/test/jsonp_spec.ts:54:9 <- angular/packages/common/http/test/jsonp_spec.js:60:21
	    at ZoneDelegate.invoke (packages/zone.js/bundles/zone.umd.js:395:30)
```

JiaLiPassion replied:
Got it, thanks for the confirmation.

JiaLiPassion commented on packages/common/http/src/jsonp.ts:
```diff
@@ -189,9 +189,12 @@ export class JsonpClientBackend implements HttpBackend {
-      node.addEventListener('error', onError);
+      // and add it to the page. We wrap it in an extra Promise, to ensure the microtask
+      // is scheduled after the loaded endpoint has executed any potential microtask itself,
+      // which is not guaranteed in Internet Explorer and EdgeHTML.
```
Could you also add the issue link https://github.com/angular/angular/issues/39496 here?

sebastianhaeni replied:
sure. also added it to the test.

JiaLiPassion commented on packages/common/http/src/jsonp.ts:
```diff
@@ -50,6 +50,8 @@ export abstract class JsonpCallbackContext {
  */
 @Injectable()
 export class JsonpClientBackend implements HttpBackend {
+  private resolvedPromise?: Promise<void>;
```
To reduce the bundle size, I think we can just initialize the `resolvedPromise` here like
```
private resolvedPromise = Promise.resolve();
```

so we can just use it directly without calling a `lazy check function`, what do you think?

sebastianhaeni replied:
yes, sure. Updated it.

JiaLiPassion reviewed (approved):
LGTM, thank you for the research and the fix!

mhevery reviewed (approved):
reviewed-for: global-approvers

mhevery commented:
[presubmit](http://test/OCL:342714141:BASE:342714166:1605563871777:4ff938d7)
